### PR TITLE
Implement feature flagged publish command

### PR DESCRIPTION
**Pull request by billlevine**

### CODE REVIEW

**Review by zmitchell — COMMENTED** [review]
@mkenigs is going to give this a more thorough review

File: cli/flox/src/commands/publish.rs
Comment (mkenigs) [inline cli/flox/src/commands/publish.rs:78]: nonblocking question: would it be better for testing if this was inside `publish_provider.publish`? Not sure, just wondering

File: cli/flox-rust-sdk/src/providers/publish.rs
Comment (mkenigs) [inline cli/flox-rust-sdk/src/providers/publish.rs:429]: suggestion followup: I think to actually test this we should make an assertion about what `publish_build` gets called with, which would require some changes to the mock client.

I think for the catalog it was problematic a few times that we did mocking at the catalog provider level rather than at the `catalog_api_v1` level. Not sure what the testing plan is for publish and if `providers::catalog` is just a passthrough it's probably easiest to follow the same pattern, but if there's ever going to be any real logic in `providers::catalog` we should probably mock one level deeper

File: cli/flox-rust-sdk/src/providers/publish.rs
Comment (billlevine) [reply cli/flox-rust-sdk/src/providers/publish.rs]: It was suggested earlier (in another PR?), but I'll make another pass at error handling.  I don't fully understand the mechanics, but I think encountered errors could be mapped to more context dependent publish errors.

File: cli/flox/src/commands/publish.rs
Comment (billlevine) [reply cli/flox/src/commands/publish.rs:33]: I don't think so... mainly copypasta.

File: cli/flox/src/commands/publish.rs
Comment (billlevine) [reply cli/flox/src/commands/publish.rs:24]: There was a conflict b/c the `Publish` trait is also in scope.  But for builds the trait is named with `er` so I updated the `Publish` trait to be `Publisher` to resolve the conflict.

File: cli/flox/src/commands/publish.rs
Comment (billlevine) [reply cli/flox/src/commands/publish.rs:80]: Noted for follow up.

File: cli/flox/src/commands/publish.rs
Comment (billlevine) [reply cli/flox/src/commands/publish.rs:39]: I guess I don't have a strong opinion one way or another (ignoring consistency considerations), so I guess probably?   Noted for follow up.

File: cli/flox/src/commands/publish.rs
Comment (billlevine) [reply cli/flox/src/commands/publish.rs:78]: I've got some open questions around the provider structure that may be worth a discussion with you guys.  It kinda evolved into this, butt worth another look as a follow up.

File: cli/flox-rust-sdk/src/providers/publish.rs
Comment (billlevine) [reply cli/flox-rust-sdk/src/providers/publish.rs]: Noted for follow up. Thanks.

File: cli/flox-rust-sdk/src/providers/publish.rs
Comment (billlevine) [reply cli/flox-rust-sdk/src/providers/publish.rs:30]: Made a pass to clean this up a bit by formatting messaging.  Will probably need a once over for error handling once the flow is nailed down for a good UX.

File: cli/flox/src/commands/publish.rs
Comment (dcarley) [inline cli/flox/src/commands/publish.rs:22]: Don't need this since it's already plumbed into `commands::InternalCommands`.
```suggestion
```

File: cli/flox/src/commands/publish.rs
Comment (dcarley) [inline cli/flox/src/commands/publish.rs:98]: (non-blocking) Should never hit this, but just in case the behaviour of `ensure_floxhub_token` changes. Sidenote: maybe it should actually return the token?
```suggestion
        let token = flox
            .floxhub_token
            .as_ref()
            .expect("should be authenticated to FloxHub");
```

File: cli/flox-rust-sdk/src/providers/publish.rs
Comment (dcarley) [inline cli/flox-rust-sdk/src/providers/publish.rs:6]: (non-blocking, same for `providers::catalog`) Should these be `debug`? Trace is pretty noisy.

File: cli/flox-rust-sdk/src/providers/publish.rs
Comment (dcarley) [inline cli/flox-rust-sdk/src/providers/publish.rs:246]: Can this be `Result<LockedUrlInfo, PublishError>` rather than `Result<Option<LockedUrlInfo>, PublishError>` now that we either return something or an error, until we allow no base catalog dependency again?

This part drew my attention to it because I was wondering why it would ever be `None` after `check_environment_metadata()`: https://github.com/flox/flox/pull/2324/files#diff-afe9925d59eecc04a1c6a9b929a6fc7861bb67e2cd20edb486f93a9661f24c73R152-R157

**Review by dcarley — APPROVED** [review]
Looks like previous comments have been addressed but will leave it up to you whether you want a re-review from one of the original reviewers.

File: cli/flox/src/commands/publish.rs
Comment (billlevine) [reply cli/flox/src/commands/publish.rs:33]: Yeah, forgot to clean that up.  Thanks.

File: cli/flox/src/commands/publish.rs
Comment (billlevine) [reply cli/flox/src/commands/publish.rs:98]: Good idea.  I have a few follow up items for later tickets (like also using the `create_test_token` elsewhere).  Added it to that list and took your suggestion.

File: cli/flox-rust-sdk/src/providers/publish.rs
Comment (billlevine) [reply cli/flox-rust-sdk/src/providers/publish.rs:246]: Probably a good idea!

**Comment by billlevine** [comment]
Do NOT merge (yet).  We want to get the service side in and have that update the client first, then rebase on that.

**Pull request merged**
